aes/src/x86/vaes512/aes256.rs

```diff
@@ -0,0 +1,687 @@
+        "vaesenc zmm16, zmm16, zmm12",
+        "vaesenc zmm17, zmm17, zmm12",
+        "vaesenc zmm18, zmm18, zmm12",
+        "vaesenc zmm19, zmm19, zmm12",
+        "vaesenc zmm20, zmm20, zmm12",
+        "vaesenc zmm21, zmm21, zmm12",
+        "vaesenc zmm22, zmm22, zmm12",
+        "vaesenc zmm23, zmm23, zmm12",
+        "vaesenc zmm24, zmm24, zmm12",
+        "vaesenc zmm25, zmm25, zmm12",
+        "vaesenc zmm26, zmm26, zmm12",
+        "vaesenc zmm27, zmm27, zmm12",
+        "vaesenc zmm28, zmm28, zmm12",
+        "vaesenc zmm29, zmm29, zmm12",
+        "vaesenc zmm30, zmm30, zmm12",
+        "vaesenc zmm31, zmm31, zmm12",
+        // aes-256 round 13 encrypt
+        "vaesenc zmm16, zmm16, zmm13",
+        "vaesenc zmm17, zmm17, zmm13",
+        "vaesenc zmm18, zmm18, zmm13",
+        "vaesenc zmm19, zmm19, zmm13",
+        "vaesenc zmm20, zmm20, zmm13",
+        "vaesenc zmm21, zmm21, zmm13",
+        "vaesenc zmm22, zmm22, zmm13",
+        "vaesenc zmm23, zmm23, zmm13",
+        "vaesenc zmm24, zmm24, zmm13",
+        "vaesenc zmm25, zmm25, zmm13",
+        "vaesenc zmm26, zmm26, zmm13",
+        "vaesenc zmm27, zmm27, zmm13",
+        "vaesenc zmm28, zmm28, zmm13",
+        "vaesenc zmm29, zmm29, zmm13",
+        "vaesenc zmm30, zmm30, zmm13",
+        "vaesenc zmm31, zmm31, zmm13",
+        // aes-256 round 14 encrypt
+        "vaesenclast zmm16, zmm16, zmm14",
+        "vaesenclast zmm17, zmm17, zmm14",
+        "vaesenclast zmm18, zmm18, zmm14",
+        "vaesenclast zmm19, zmm19, zmm14",
+        "vaesenclast zmm20, zmm20, zmm14",
+        "vaesenclast zmm21, zmm21, zmm14",
+        "vaesenclast zmm22, zmm22, zmm14",
+        "vaesenclast zmm23, zmm23, zmm14",
+        "vaesenclast zmm24, zmm24, zmm14",
+        "vaesenclast zmm25, zmm25, zmm14",
+        "vaesenclast zmm26, zmm26, zmm14",
+        "vaesenclast zmm27, zmm27, zmm14",
+        "vaesenclast zmm28, zmm28, zmm14",
+        "vaesenclast zmm29, zmm29, zmm14",
+        "vaesenclast zmm30, zmm30, zmm14",
+        "vaesenclast zmm31, zmm31, zmm14",
+        // save cipher-data
+        "vmovdqu32 [{optr} +  0 * 64], zmm16",
+        "vmovdqu32 [{optr} +  1 * 64], zmm17",
+        "vmovdqu32 [{optr} +  2 * 64], zmm18",
+        "vmovdqu32 [{optr} +  3 * 64], zmm19",
+        "vmovdqu32 [{optr} +  4 * 64], zmm20",
+        "vmovdqu32 [{optr} +  5 * 64], zmm21",
+        "vmovdqu32 [{optr} +  6 * 64], zmm22",
+        "vmovdqu32 [{optr} +  7 * 64], zmm23",
+        "vmovdqu32 [{optr} +  8 * 64], zmm24",
+        "vmovdqu32 [{optr} +  9 * 64], zmm25",
+        "vmovdqu32 [{optr} + 10 * 64], zmm26",
+        "vmovdqu32 [{optr} + 11 * 64], zmm27",
+        "vmovdqu32 [{optr} + 12 * 64], zmm28",
+        "vmovdqu32 [{optr} + 13 * 64], zmm29",
+        "vmovdqu32 [{optr} + 14 * 64], zmm30",
+        "vmovdqu32 [{optr} + 15 * 64], zmm31",
+
+        keys = in(reg) keys.as_ptr(),
+        iptr = in(reg) iptr,
+        optr = in(reg) optr,
+
+        clobber_abi("C"),
```

Comment:
IIUC you use it to clobber the ZMM registers, correct? But `clobber_abi("C")` has [no effect](https://doc.rust-lang.org/reference/inline-assembly.html#abi-clobbers) on them. It looks like compiler accepts `out("zmm0") _`, but I am not sure if it has the desired effect.

Reply:
> IIUC you use it to clobber the ZMM registers, correct? But `clobber_abi("C")` has [no effect](https://doc.rust-lang.org/reference/inline-assembly.html#abi-clobbers) on them. It looks like compiler accepts `out("zmm0") _`, but I am not sure if it has the desired effect.

I was under the impression that this was an omission (and these are indeed included in the C ABI clobbers) due to the fact that the `zmm` registers are aliased (i.e., that it would be redundant to list them there since `xmm` are already listed):

https://doc.rust-lang.org/reference/inline-assembly.html#register-names

Do you believe that is incorrect?

> It's also probably worth to add the other flags like `nostack` and `preserve_flags`. It would be nice to also use pure, but it would require passing of keys and block data in input/output arguments of asm! blocks.

I do use the `nostack` and `preserve_flags` there (on the next line).

I don't remember for sure now but I believe I may have tried using `pure` but opted not to keep it for some reason. I'll give it another try.

Reply:
>Do you believe that is incorrect?

I don't know, but I would prefer to explicitly clobber only used registers. `clobber_abi("C")` will clobber other registers which we do not use.

>I do use the nostack and preserve_flags there (on the next line).

Ah, true.

>I don't remember for sure now but I believe I may have tried using pure but opted not to keep it for some reason.

It requires to be used together with either `nomem` or `readonly`, meaning that you can not do write of blocks data in inline asm. It should be possible to do for the vaes256 backends, but not for vaes512.

Reply:
I've added the explicit clobbers and `pure` and `readonly` for the VAES 256 case.

The only thing preventing us from doing that for the VAES 512 case is that the `asm!` macro won't let you use a `zmm` register as an output without the `avx512f` feature. Presumably because it needs to use some instruction equivalent to an intrinsic which is still unstable, because the actual `__m512i` type is stabilized already.